### ISSUE TITLE
Extract function assist misses bindings only used in the else branch of let-else

### ISSUE THREAD

**Issue by svix-jplatte** [rust-lang/rust-analyzer#21039]
The extract function assist fails to add parameters for bindings only used in let-eles to the generated function. Reproducer below (generated function should have `x` as a parameter, but has none).

**rust-analyzer version**: 0.4.2679-standalone
**rustc version**: 1.93.0-nightly (01867557c 2025-11-12)
**editor or extension**: VSCode
**code snippet to reproduce**:
```rust
fn foo() -> u32 {
    let x = 5;

    // extract this let-else branch into a new function
    let Some(y) = Some(1) else {
        return x * 2;
    };

    y
}
```

**Comment by Natural-selection1** [comment]
let-else requires that the else block must always diverge (evaluate to the [never type](https://doc.rust-lang.org/reference/types/never.html)). 

In the vast majority of scenarios, this conflicts with the "Extract function" operation, such as anything related to control flow (`break, continue, return, ?`). 
Only in cases like `unreachable!()` or `panic!()` would "Extract function" make sense. 

So I would say that your requirement itself is not entirely reasonable. However, we should indeed avoid displaying the "Extract function" option when control flow that goes beyond the selected scope is detected within the text selection.

**Comment by svix-jplatte** [comment]
That same argument applies to any code with early returns, no? Which rust-analyzer handles just fine (sometimes introducing a `ControlFlow` that I usually refactor away after, but it's still very useful). In the real-world case where I stumbled upon this, the assist definitely helped even though it missed two parameters and got the return type slightly wrong.

**Comment by dfireBird** [comment]
I'll pick this up since I had just done with an extract function just an day back (If no one is working that is?)

@rustbot claim

**Comment by svix-jplatte** [comment]
Thanks a bunch @dfireBird and @ChayimFriedman2! :)
